### PR TITLE
power: stm32l5: support of the Low Power sleep mode

### PR DESCRIPTION
**Pull request by FRASTM**
With this patch, the low power mode on stm32L5 is enabling the Low Power Timer to measure the sleep time duration and wakeup from sleep mode on this LP timer expiration.
The SLEEP modes (corresponding to STM32 stop 0,1,2 modes) are supported on this device.
When low power is enabled on the system and when going to sleep, the lptim counts and wakes up the kernel though its interrupt routine.
The soc policy decides to go to sleep, if compatible with its residency when calling the sys_pm_policy_next_state for a nb of kernel ticks.

This PR is requires the support of the LPTimer on stm32l5  #32999

Note that DEEP_SLEEP modes are not enabled yet, as LPTIM cannot run during stm32 standby or shutdown modes.

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

**Comment by FRASTM** [comment]
Merge lptim an PM is an single PR
Rebase
Set the LSE clock as default source clock for stm32l5xx boards. Consequently the LSE oscillator propagation for system clock must be  enabled for this soc serie. Setting the LSE system clock (LSESYS) bit of the RCC BDCR register.
Add a fix () to avoid multiple declaration of `z_clock_set_timeout` etc. when the STM32_LPTIM_TIMER is configured (not compatible with CORTEX_M_SYSTICK)

**Comment by FRASTM** [comment]
Rebasing with the definition of the CORTEX_M_SYSTICK
LPTIM source clock are set to LSE

File: soc/arm/st_stm32/stm32l5/power.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019 STMicroelectronics.
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/power.c:2]: 2021

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series [hunk 2]
```diff
@@ -10,4 +10,14 @@ source "soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l5*"
 config SOC_SERIES
 	default "stm32l5"
 
+if PM
+config PM_DEVICE
+	bool
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series:15]: Remove bool

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series [hunk 3]
```diff
@@ -10,4 +10,14 @@ source "soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l5*"
...
+	default y
+
+config STM32_LPTIM_TIMER
+	bool
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series:19]: Remove bool

File: soc/arm/st_stm32/stm32l5/power.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+/* Initialize STM32 Power */
+static int stm32_power_init(const struct device *dev)
+{
+	unsigned int ret;
+
+	ARG_UNUSED(dev);
+
+	ret = irq_lock();
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/power.c:128]: I don't think this is required

File: soc/arm/st_stm32/stm32l5/power.c [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
+		LL_LPM_EnableDeepSleep();
+		k_cpu_idle();
+		break;
+	case 2:
+		/* this corresponds to the STOP2 mode: */
+#ifdef CONFIG_DEBUG
+		/* Enable the Debug Module during STOP mode */
+		LL_DBGMCU_EnableDBGStopMode();
+#endif /* CONFIG_DEBUG */
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l5/power.c:71]: I'm not sure we need to do this at this step.
So maybe do it once for all at init.

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32l5/power.c (hunk 4 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/power.c:128]: removed

File: soc/arm/st_stm32/stm32l5/power.c (hunk 5 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/power.c:71]: yes, put in the `stm32_power_init() `
Note this is valid for all the low power modes : STOP 0/1/2 and to be completed with LL_DBGMCU_EnableDBGSleepMode if needed.

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series (hunk 2 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series]: ok

File: soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series (hunk 3 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series]: ok

**Review by erwango — APPROVED** [review]

**Comment by FRASTM** [comment]
At this time, only **stm32l562e_dk** board has low power sleep modes in the stm32l5xx soc serie.

**Comment by FRASTM** [comment]
rebase , fixing conflicts with stm32l562e_dk.dts

**Comment by FRASTM** [comment]
adding a commit to enable the MSI auto-calibration on the disco board

File: boards/arm/stm32l562e_dk/stm32l562e_dk.dts [hunk 6]
```diff
@@ -21,12 +21,37 @@
...
+			substate-id = <1>;
+			min-residency-us = <500>;
+		};
+		stop2: state2 {
+			compatible = "zephyr,power-state";
+			power-state-name = "suspend-to-idle";
+			substate-id = <2>;
+			min-residency-us = <900>;
+		};
+	};
```
Comment (erwango) [inline boards/arm/stm32l562e_dk/stm32l562e_dk.dts:43]: Please move this to `boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi`

Comment (FRASTM) [reply boards/arm/stm32l562e_dk/stm32l562e_dk.dts]: ok, moved to boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi

**Comment by FRASTM** [comment]
- moving power state definition to boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi
- adding the lptim1 and low power modes to the boards/arm/stm32l562e_dk_ns 
- change the substate-id range from 1-3 as described in the _include/power/power_state.h_
 this is changing the range in `pm_power_state_set() `of the soc/arm/st_stm32/stm32l5/power.c

File: soc/arm/st_stm32/stm32l5/power.c [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+/* Handle SOC specific activity after Low Power Mode Exit */
+void pm_power_state_exit_post_ops(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power substate-id %u", info.state);
+	} else {
+		switch (info.substate_id) {
+		case 0:	/* STOP0 */
+			__fallthrough;
+		case 1:	/* STOP1 */
```
Comment (ABOSTM) [inline soc/arm/st_stm32/stm32l5/power.c:84]: Suported substate_id are 1, 2, 3.
substate_id = 1 correspond to STOP0 in dtsi file

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Commit Message:
```
The low power states 0∕1∕2 are added to the stm32l562e_dk board
with (default min residency values).
```
Sentence not finished: With what ? (or wrong parenthesis)

**Comment by FRASTM** [comment]
rebase

File: boards/arm/stm32l562e_dk/stm32l562e_dk_defconfig [hunk 8]
```diff
@@ -21,6 +21,7 @@ CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
 # PLL configuration
 CONFIG_CLOCK_STM32_PLL_SRC_MSI=y
 CONFIG_CLOCK_STM32_MSI_RANGE=6
+CONFIG_CLOCK_STM32_MSI_PLL_MODE=y
```
Comment (erwango) [inline boards/arm/stm32l562e_dk/stm32l562e_dk_defconfig:24]: This symbol has a dependency on CONFIG_CLOCK_STM32_LSE=y. It should be added

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32l5/power.c (hunk 7 above)
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l5/power.c]: correct

**Review by ABOSTM — APPROVED** [review]
LGTM

File: boards/arm/stm32l562e_dk/stm32l562e_dk_defconfig (hunk 8 above)
Comment (FRASTM) [reply boards/arm/stm32l562e_dk/stm32l562e_dk_defconfig:24]: could be 
```
config CLOCK_STM32_MSI_PLL_MODE
	default y
	depends on STM32_LPTIM_CLOCK_LSE
```
in the soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series

**Comment by FRASTM** [comment]
rebase
add the CONFIG_CLOCK_STM32_LSE  in the board def_config to enable the CONFIG_CLOCK_STM32_MSI_PLL_MODE without warning (as the CLOCK_STM32_MSI_PLL_MODE depends on the CLOCK_STM32_LSE through the Kconfig.stm32l4_l5_wb_wl)

**Review by erwango — APPROVED** [review]

**Pull request merged**